### PR TITLE
create forgot_password screen

### PR DESCRIPTION
**Pull request by vyvanhungbg**
## Related Tickets
- [#Task #54662: Create layout xml](https://edu-redmine.sun-asterisk.vn/issues/54662)
- [#Task #54663: Create login , call api](https://edu-redmine.sun-asterisk.vn/issues/54663)
## WHAT
- Create a feature screen Forgot password
## Evidence (Screenshot or Video)
image

## Review Checklist

Category | View Point | Description | Self review | Reviewer2 (name)
--- | --- | --- | --- | ---
Conventions | Sun* coding conventions | https://sal.vn/WtYPAy |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Conventions | Kotlin coding conventions | https://kotlinlang.org/docs/coding-conventions.html |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Redmine | Sun* Redmine working process  | https://sal.vn/piPoQn |<li>- [ ] yes</li>|<li>- [ ] yes</li>  

## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

**Review by cuong117 — APPROVED** [review]

**Review by ndt0409 — APPROVED** [review]

File: app/src/main/java/com/atom/android/bookshop/data/source/remote/forgotpassword/ForgotPasswordDataSource.kt [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import com.atom.android.bookshop.data.source.remote.api.ApiURL
+import com.atom.android.bookshop.data.source.remote.api.convertJsonToResponseObject
+import com.atom.android.bookshop.utils.JSONConvertException
+import com.atom.android.bookshop.utils.handler
+import com.atom.android.bookshop.utils.httpConnectionSendFormData
+import com.atom.android.bookshop.utils.remoteExecuteCallAPI
+import org.json.JSONException
+import org.json.JSONObject
+
+class ForgotPasswordDataSource : IForgotPasswordDataSource.Remote {
```
Comment (vietnh-1109) [inline app/src/main/java/com/atom/android/bookshop/data/source/remote/forgotpassword/ForgotPasswordDataSource.kt]:
```suggestion
class ForgotPasswordRemoteDataSource : IForgotPasswordDataSource.Remote {
```

File: app/src/main/java/com/atom/android/bookshop/data/source/remote/forgotpassword/ForgotPasswordDataSource.kt [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        val jsonObjectResponseObject = JSONObject(
+            httpConnectionSendFormData(
+                dataFormEmail,
+                ApiURL.pathForgotEmail(),
+                ApiConstants.Method.POST
+            )
+        )
+        try {
+            val responseObject = convertJsonToResponseObject(jsonObjectResponseObject)
+            if (responseObject.status) {
```
Comment (vietnh-1109) [inline app/src/main/java/com/atom/android/bookshop/data/source/remote/forgotpassword/ForgotPasswordDataSource.kt]:
```suggestion
            if (responseObject.isSuccess) {
```

File: app/src/main/java/com/atom/android/bookshop/ui/authentication/forgotpassword/ForgotPasswordFragment.kt [hunk 3]
```diff
@@ -21,11 +68,15 @@ class ForgotPasswordFragment :
         // late impl
     }
 
-
     override fun initEvent() {
         binding.textViewLogin.setOnClickListener {
             navigate(R.id.action_navigation_forgot_password_to_navigation_login)
         }
+        binding.btnForgotPassword.setOnClickListener {
+            progessBar.start(true)
```
Comment (vietnh-1109) [inline app/src/main/java/com/atom/android/bookshop/ui/authentication/forgotpassword/ForgotPasswordFragment.kt]: progress bar viết trong presenter chứ

File: app/src/main/java/com/atom/android/bookshop/ui/authentication/forgotpassword/ForgotPasswordFragment.kt [hunk 4]
```diff
@@ -39,4 +90,20 @@ class ForgotPasswordFragment :
         }
     }
 
+    override fun requestSuccess(message: String?) {
+        countDownTimer?.start()
+        progessBar.dismiss()
```
Comment (vietnh-1109) [inline app/src/main/java/com/atom/android/bookshop/ui/authentication/forgotpassword/ForgotPasswordFragment.kt]: tương tự nhé , listener ra view xong mới dismiss hoặc show

File: app/src/main/java/com/atom/android/bookshop/ui/authentication/forgotpassword/ForgotPasswordPresenter.kt [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        // late impl
+    }
+
+    companion object {
+        private var instance: ForgotPasswordPresenter? = null
+        fun getInstance(
+            repository: ForgotPasswordRepository,
+            forgotPasswordView: ForgotPasswordContract.View
+        ) =
+            synchronized(this) {
```
Comment (vietnh-1109) [inline app/src/main/java/com/atom/android/bookshop/ui/authentication/forgotpassword/ForgotPasswordPresenter.kt]:
```suggestion
        ) = synchronized(this) {
```

File: app/src/main/java/com/atom/android/bookshop/utils/ToastUtils.kt [hunk 6]
```diff
@@ -1,7 +1,13 @@
...
+import android.text.Spannable
+import android.text.SpannableString
+import android.text.style.ForegroundColorSpan
+import android.text.style.RelativeSizeSpan
 import android.widget.Toast
+import androidx.core.content.ContextCompat
+import com.atom.android.bookshop.R
 
 fun Context.toast(message: String?) {
     Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
```
Comment (vietnh-1109) [inline app/src/main/java/com/atom/android/bookshop/utils/ToastUtils.kt]: này là extension , thì phải đặt là ContextExtension chứ

File: app/src/main/java/com/atom/android/bookshop/utils/ToastUtils.kt [hunk 7]
```diff
@@ -10,3 +16,39 @@ fun Context.toast(message: String?) {
 fun Context.toast(resource: Int) {
     Toast.makeText(this, resource, Toast.LENGTH_SHORT).show()
 }
+
+fun Context.textSpannable(text: String?, number: String): Spannable {
+    val spannable: Spannable = SpannableString(text + number)
```
Comment (vietnh-1109) [inline app/src/main/java/com/atom/android/bookshop/utils/ToastUtils.kt]:
```suggestion
    val spannable = SpannableString(text + number)
```

**Pull request merged**